### PR TITLE
added JobInvocation and JobTemplate entities

### PR DESCRIPTION
**Pull request by pondrejk**
Added basic entities for job templates and job invocations, related to #400 

```
In [4]: tmp = entities.JobTemplate(id=121)
In [5]: tmp.read_json()
Out[5]:
{u'audit_comment': None,
 u'created_at': u'2017-04-11 11:18:35 UTC',
 u'description_format': None,
 u'effective_user': {u'current_user': False,
  u'overridable': True,
  u'value': None},
 u'id': 121,
 u'job_category': u'Miscellaneous',
 u'locations': [],
 u'locked': False,
 u'name': u'wvbuPgW',
 u'organizations': [{u'description': None,
   u'id': 31,
   u'name': u'oPIrBo',
   u'title': u'oPIrBo'}],
 u'provider_type': u'SSH',
 u'snippet': False,
 u'template': u'getenforce\n',
 u'template_inputs': [],
 u'updated_at': u'2017-04-11 11:18:35 UTC'}
```
```
In [6]: inv = entities.JobInvocation(id=20)
In [7]: inv.read_json()
Out[7]: 
{u'description': u'Run ls',
 u'dynflow_task': {u'id': u'5fdc935f-c83a-46e0-8c99-5ed61a103194',
  u'state': u'stopped'},
 u'failed': 0,
 u'id': 20,
 u'job_category': u'Commands',
 u'pending': 0,
 u'start_at': u'2017-04-12 09:36:52 UTC',
 u'status': 0,
 u'status_label': u'succeeded',
 u'succeeded': 1,
 u'targeting': {u'bookmark_id': None,
  u'hosts': [{u'id': 1, u'name': u'xxx'}],
  u'search_query': u'name ~ xxx or name ~ xxx',
  u'targeting_type': u'static_query',
  u'user_id': 3},
 u'targeting_id': 20,
 u'task': {u'id': u'5fdc935f-c83a-46e0-8c99-5ed61a103194',
  u'state': u'stopped'},
 u'template_invocations': [{u'template_id': 97,
   u'template_invocation_input_values': [{u'template_input_id': 7,
     u'template_input_name': u'command',
     u'value': u'ls'}],
   u'template_name': u'Run Command - SSH Default'}],
 u'total': 1}

```

### CODE REVIEW

File: nailgun/entities.py [hunk 1]
```diff
@@ -3270,6 +3270,72 @@ def search_normalize(self, results):
         return super(Interface, self).search_normalize(results)
 
 
+class JobInvocation(
+        Entity,
```
Comment (svtkachenko) [inline nailgun/entities.py]: In the docs I can't find the way to Update/Delete this entity, only Create/Read. If so, please remove unneeded mixins.

File: nailgun/entities.py [hunk 2]
```diff
@@ -3270,6 +3270,72 @@ def search_normalize(self, results):
...
+            'created_at': entity_fields.DateTimeField(),
+            'updated_at': entity_fields.DateTimeField(),
+            'template': entity_fields.StringField(),
+            'locked': entity_fields.BooleanField(),
+            'job_category': entity_fields.StringField(),
+            'provider_type': entity_fields.StringField(),
+            'snippet': entity_fields.BooleanField(),
+            'template_inputs': entity_fields.ListField(),
+            'effective_user': entity_fields.DictField(),
+            'locations': entity_fields.ListField(),
```
Comment (svtkachenko) [inline nailgun/entities.py]:
```python
'location': entity_fields.OneToManyField(Location),
```

File: nailgun/entities.py [hunk 3]
```diff
@@ -3270,6 +3270,72 @@ def search_normalize(self, results):
...
+            'organizations': entity_fields.ListField(),
```
Comment (svtkachenko) [inline nailgun/entities.py]:
```python
'organization': entity_fields.OneToManyField(Organization),
```

**Comment by svtkachenko** [comment]
@pondrejk Can you please also provide results for `read()` and not only `read_json()`?

**Comment by pondrejk** [comment]
Thanks @svtkachenko, comments addressed + read() output:

```
In [4]: inv = entities.JobInvocation(id=18)

In [5]: inv.read()
2017-04-25 16:51:19 - nailgun.client - DEBUG - Making HTTP GET request to .../api/v2/job_invocations/18 with options {'verify': False, 'auth': ('admin', 'changeme'), 'headers': {'content-type': 'application/json'}} and no data.
2017-04-25 16:51:20 - nailgun.client - DEBUG - Received HTTP 200 response: {"id":18,"description":"Run ls","job_category":"Commands","targeting_id":18,"status":0,"start_at":"2017-04-12 09:29:37 UTC","status_label":"succeeded","dynflow_task":{"id":"a171f17f-bf9b-4b94-8db9-6508d0a78a2a","state":"stopped"},"succeeded":1,"failed":0,"pending":0,"total":1,"targeting":{"bookmark_id":null,"search_query":"name ~ ... ~ ...","targeting_type":"static_query","user_id":3,"hosts":[{"name":"...,"id":1}]},"task":{"id":"a171f17f-bf9b-4b94-8db9-6508d0a78a2a","state":"stopped"},"template_invocations":[{"template_id":97,"template_name":"Run Command - SSH Default","template_invocation_input_values":[{"template_input_name":"command","template_input_id":7,"value":"ls"}]}]}
Out[5]: nailgun.entities.JobInvocation(status=0, start_at=u'2017-04-12 09:29:37 UTC', description=u'Run ls', targeting_id=18, total=1, id=18, task={u'state': u'stopped', u'id': u'a171f17f-bf9b-4b94-8db9-6508d0a78a2a'}, targeting={u'search_query': u'name ~ ... or name ~ ...', u'bookmark_id': None, u'user_id': 3, u'hosts': [{u'name': u'...', u'id': 1}], u'targeting_type': u'static_query'}, succeeded=1, status_label=u'succeeded', dynflow_task={u'state': u'stopped', u'id': u'a171f17f-bf9b-4b94-8db9-6508d0a78a2a'}, failed=0, template_invocations=[{u'template_name': u'Run Command - SSH Default', u'template_id': 97, u'template_invocation_input_values': [{u'template_input_id': 7, u'template_input_name': u'command', u'value': u'ls'}]}], job_category=u'Commands', pending=0)

In [6]: tmp = entities.JobTemplate(id=121)

In [7]: tmp.read()
2017-04-25 16:51:37 - nailgun.client - DEBUG - Making HTTP GET request to https://.../api/v2/job_templates/121 with options {'verify': False, 'auth': ('admin', 'changeme'), 'headers': {'content-type': 'application/json'}} and no data.
2017-04-25 16:51:38 - nailgun.client - DEBUG - Received HTTP 200 response: {"audit_comment":null,"description_format":null,"created_at":"2017-04-11 11:18:35 UTC","updated_at":"2017-04-11 11:18:35 UTC","template":"getenforce\n","locked":false,"id":121,"name":"wvbuPgW","job_category":"Miscellaneous","provider_type":"SSH","snippet":false,"template_inputs":[],"effective_user":{"value":null,"current_user":false,"overridable":true},"locations":[],"organizations":[{"id":31,"name":"oPIrBo","title":"oPIrBo","description":null}]}
Out[7]: nailgun.entities.JobTemplate(description_format=None, updated_at=u'2017-04-11 11:18:35 UTC', effective_user={u'overridable': True, u'value': None, u'current_user': False}, name=u'wvbuPgW', template_inputs=[], locked=False, created_at=u'2017-04-11 11:18:35 UTC', id=121, snippet=False, location=[], template=u'getenforce\n', job_category=u'Miscellaneous', organization=[nailgun.entities.Organization(id=31)], audit_comment=None, provider_type=u'SSH')

```

File: nailgun/entities.py [hunk 4]
```diff
@@ -3270,6 +3270,68 @@ def search_normalize(self, results):
...
+
+    def __init__(self, server_config=None, **kwargs):
+        self._fields = {
+            'description': entity_fields.StringField(),
+            'job_category': entity_fields.StringField(),
+            'targeting_id': entity_fields.IntegerField(),
+            'status': entity_fields.IntegerField(),
+            'start_at': entity_fields.DateTimeField(),
+            'status_label': entity_fields.StringField(),
+            'dynflow_task': entity_fields.ListField(),
```
Comment (abalakh) [inline nailgun/entities.py]: Isn't it ForemanTask? Should be `'dynflow_task': OneToOneField(ForemanTask)` then. And according to your console output it's a dict, not a list anyways.

File: nailgun/entities.py [hunk 5]
```diff
@@ -3270,6 +3270,68 @@ def search_normalize(self, results):
...
+            'succeeded': entity_fields.IntegerField(),
+            'failed': entity_fields.IntegerField(),
+            'pending': entity_fields.IntegerField(),
+            'total': entity_fields.IntegerField(),
+            'targeting': entity_fields.DictField(),
+            'task': entity_fields.DictField(),
```
Comment (abalakh) [inline nailgun/entities.py]: `OneToOneField(ForemanTask)` maybe?

File: nailgun/entities.py [hunk 6]
```diff
@@ -3270,6 +3270,68 @@ def search_normalize(self, results):
...
+            'targeting_id': entity_fields.IntegerField(),
+            'status': entity_fields.IntegerField(),
+            'start_at': entity_fields.DateTimeField(),
+            'status_label': entity_fields.StringField(),
+            'dynflow_task': entity_fields.ListField(),
+            'succeeded': entity_fields.IntegerField(),
+            'failed': entity_fields.IntegerField(),
+            'pending': entity_fields.IntegerField(),
+            'total': entity_fields.IntegerField(),
+            'targeting': entity_fields.DictField(),
```
Comment (abalakh) [inline nailgun/entities.py]: `targeting_id` field in console output looks very suspicious to me. Most likely it should be some entity, not raw output. Could you please double check?

File: nailgun/entities.py [hunk 7]
```diff
@@ -3270,6 +3270,68 @@ def search_normalize(self, results):
...
+            'task': entity_fields.DictField(),
+            'template_invocations': entity_fields.ListField(),
```
Comment (abalakh) [inline nailgun/entities.py]: Same here - seems like it should be a list of entities (OneToManyField).

File: nailgun/entities.py [hunk 8]
```diff
@@ -3270,6 +3270,68 @@ def search_normalize(self, results):
...
+        EntityCreateMixin,
+        EntityDeleteMixin,
+        EntityReadMixin,
+        EntitySearchMixin,
+        EntityUpdateMixin):
+    """A representation of a Job invocation entity."""
+
+    def __init__(self, server_config=None, **kwargs):
+        self._fields = {
+            'name': entity_fields.StringField(),
```
Comment (abalakh) [inline nailgun/entities.py]: Please sort fields alphabetically.

**Comment by pondrejk** [comment]
Thanks guys for reviews, I applied the comments but I figured out there is more work to be done on job invocation create, therefore putting do not merge for now

**Comment by san7ket** [comment]
@pondrejk is this still valid?

**Comment by oshtaier** [comment]
JobTemplate entity is already covered
Cover JobInvocation in #596

Took part of the code from here and re-tested. Thanks for ideas
Closing as nobody looks here

**Pull request closed without merging**